content/Building an Engine-modularized API in Rails - Guillermo Leopold.md

```diff
@@ -0,0 +1,260 @@
+
+1. Create an observer class inside our models folder in the **Payment engine** as we want to trigger an action in this engine as a consequence of an action in the **User** engine.
+
+2. Set up the observer in the **engine.rb** file by adding the following line:
+ ```Ruby
+ config.active_record.observers = :'payment/user_customer_observer'
+ ```
+
+3. Implement your observer, we need to specify in which model we want to observe and implement an action according that behave similar to a callback in ActiveRecord. It might look like this:
+
+```Ruby
+module Payment
+  class UserCustomerObserver < ActiveRecord::Observer
+    observe User::Customer
+    attr_reader :customer
+
+    def after_update(customer)
+      customer.update!(payments: customer.payments + 1)
+    end
+  end
+end
+```
+
+4. Finally, if you want to disable observers while running **rspec**, you can add this line in your test suite as a before action:
+```Ruby
+ActiveRecord::Base.observers.disable :all
+```
+
+On the other hand, testing inside an engine with a rspec it’s quite similar to what we are used to doing and it provides us with some guarantees and the only considerations that took us some time are:
+  * We will need to load manually factories from other engines when using FactoryBot in our engine
+  * We need to require in our rails_helper.rb any development dependency that we are using in our test suite
+
+After developing a feature, one of the advantages that we found out about working with engines is that we can just run the test suite for the engine where we have been working on without the necessity to go through all the suite tests in the application. This reduces the time to run tests locally dramatically.
+
+The API architecture and the project structure will look something like this:
+
+![Engines Structure + Folder Struct](images/engines_structure_folder_struct.png)
+
+## Summing up
+
+It’s been a while since we started working with this project structure. Right now we are at a point where we can already look back and find the pros and cons of making this particular decision:
+
+**Pros:**
+  * Helps to parallelize work when working with medium-sized development teams.
+  * Makes the code more organized and easier to read or find where a feature should be implemented.
+  * The possibility to run tests inside an engine scope reduces time when testing locally.
+  * A modular application like this is a mid-step between a monolith API and microservices.
+  * Possibility to pick a full engine and use it in another API. We did this with our catalog engine and an admin that we develop in another project.
+
+
+**Cons:**
+  * Lot of configuration and changes are needed when starting working with engines, requires a bit of time during the first weeks.
+  * Lack of information on the internet, building an API with engines doesn’t seem to be a popular choice.
+  * Despite our effort to avoid repeating code, we still repeat more code than when working in a monolith api
```

Comment:
I would change this for: `we repeat even more code than`.

Also, small thing, the two previous paragraphs end in a `.`